js/src/data/reducer.js

```diff
@@ -37,29 +37,37 @@ const DEFAULT_STATE = {
 	report: {},
 };
 
-const getNextStateForShipping = ( state ) => {
+// Referenced and modified from https://github.com/lodash/lodash/issues/1696#issuecomment-328335502
+function chain( state, basePath = '' ) {
+	const nextState = Object.assign( state.constructor(), state );
+	const customizer = ( value ) => {
+		if ( value === null || value === undefined ) {
+			return {};
+		}
+		return clone( value );
```

Comment:
What do you think if we replace this `clone` with `return Object.assign( value.constructor(), value )` so the only thing depending on lodash (in this file) is the function setWith?

Reply:
Although we will almost not have the following usage to be committed to codebase, it is inevitable that the wrong `path` might be typed when developing, and the returned results would be more confusing than in the current implementation to figure out how they are generated.

```js
function chain( state, basePath = '' ) {
	const nextState = Object.assign( state.constructor(), state );
	const customizer = ( value ) => {
		if ( value === null || value === undefined ) {
			return {};
		}
		return Object.assign( value.constructor(), value );
	};

	return {
		set( path, value ) {
			const fullPath = basePath ? `${ basePath }.${ path }` : path;
			setWith( nextState, fullPath, value, customizer );
			return this;
		},
		end: () => nextState,
	};
}

function set( state, path, value ) {
	return chain( state ).set( path, value ).end();
}

set( { foo: 123 }, 'foo.bar', 456 );
// Got { foo: [Number: 0] { bar: 456 } }
// => `foo` property is a number object with value 0,
//    and that number object has a `bar` property with 456

set( { foo: 'Hello' }, 'foo.bar', 'WordPress' );
// Got
// {
//   foo: [String: ''] {
//     '0': 'H',
//     '1': 'e',
//     '2': 'l',
//     '3': 'l',
//     '4': 'o',
//     bar: 'WordPress'
//   }
// }
// => `foo` property is a string object with enumerated keys/values of 'Hello',
//    and that string object has a `bar` property with 'WordPress'
```

The current implementation is easier to find that the wrong path was typed. Surely this is a relative edge case, so I think it won't be too many problems even if we use the proposed method instead. 

However, Sorry I didn't prepare a summary of the historical source of the solution. 🙇‍♂️  There is a lot of context to this PR.

Reducing dependency on lodash or gradually phasing out lodash is indeed a long-term goal for some teams. While we haven't explicitly decided whether to follow those principles, in the short-term goals discussed in #270, we were tending to resolve it with the effective solution found from lodash GitHub and less of our custom implementations. Some relevant comments:

- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863080575
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863169462
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-863412160
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-864708881
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-864799355
- https://github.com/woocommerce/google-listings-and-ads/issues/270#issuecomment-883266968

The goal of this PR is to solve the #270 problem based on the previous conclusion. If the concern is to reduce the reliance on lodash, I'd be happy to open another issue or p2 to discuss it because currently it's a bit out of the scope of this PR's purpose. 🙏  So, if there're no concerns other than those mentioned above, I would prefer to keep the lodash solution as it is and only adjusts the parts we can't avoid.